server/index.ts

```diff
@@ -102,36 +127,67 @@ const workflow = new StateGraph(MessagesAnnotation)
 const memory = new MemorySaver();
 const langApp = workflow.compile({ checkpointer: memory });
 
+// Track active message processing
+const activeMessages = new Map<string, AbortController>();
+
 // This is the main endpoint for use by the client app. We may want to add more, e.g. another for tool calls, etc.
 app.post("/default/davaiServer/message", async (req, res) => {
-  const { llmId, message, threadId, dataContexts } = req.body;
+  const { llmId, message, threadId, codapData, messageId } = req.body;
   const config = { configurable: { thread_id: threadId, llmId } };
 
   try {
     const messages = [];
 
-    // If we have data contexts, process them first
-    if (dataContexts && typeof dataContexts === "object") {
-      // console.log("Processing data contexts");
-      messages.push(...processDataContexts(dataContexts));
+    // If we have data contexts, process them before adding.
+    if (codapData && typeof codapData === "object") {
+      messages.push(...processCodapData(codapData));
+    } else {
+      // Add the user's message
+      messages.push({
+        role: "user",
+        content: message,
+      });
     }
 
-    // Add the user's message
-    messages.push({
-      role: "user",
-      content: message,
-    });
-
-    // console.log("Final messages array:", messages.map(m => ({ role: m.role, contentLength: m.content.length })));
-
-    const output = await langApp.invoke({ messages }, config);
+    // Create an AbortController for this request. This lets us to cancel the request if needed.
+    const controller = new AbortController();
+    activeMessages.set(messageId, controller);
+
+    const output = await langApp.invoke(
+      { messages }, 
+      { 
+        ...config,
+        signal: controller.signal 
+      }
+    );
     const lastMessage = output.messages[output.messages.length - 1];
 
+    // Clean up the controller
+    activeMessages.delete(messageId);
+
     res.json({ response: lastMessage.content });
   } catch (err: any) {
-    console.error("Error in /api/message:", err);
-    console.error("Error stack:", err.stack);
-    res.status(500).json({ error: "LangChain Error", details: err.message });
+    if (err.message === "Aborted") {
+      res.status(499).json({ error: "Message processing cancelled" });
```

Comment:
http status 499 is non-standard, not sure if that matters but it might trip up some browsers.
I'd send a 500 here too, unless you consider responding to a cancelation request to be a form of success.